### ISSUE TITLE
define equivalence on argument lists

### ISSUE THREAD

**Issue by michaelficarra** [js-choi/proposal-function-memo#5]
The most important thing for this proposal to figure out is how it determines that a subsequent call is "equivalent" to a previous call so that it can return the stored result. This will involve at least

1. Determining what comparison algorithm to apply to each positional argument.
2. Determining whether extra arguments beyond the declared parameters contribute.
3. Determining whether the `this` and `new.target` values contribute.

and probably more.

**Comment by js-choi** [comment]
My inclination is to let the developer supply any Map-like object as a cache, and to supply tuple keys to `cache.has` and `cache.get`.

**Comment by zloirock** [comment]
> `Map` would be used by default.

A strange option.

**Comment by js-choi** [comment]
@zloirock: `new Map()` would basically be the memory-unbounded version of `new LRUMap(maxNumberOfEntries)`.

**Comment by zloirock** [comment]
@js-choi anyway - I think it's good for custom cach strategy, but I think that the user should be sure that with the default cach strategy the method will not be called again with the same arguments and for arguments that GCed and no longer available the result will not pollute memore.

**Comment by js-choi** [comment]
Yes, the question of what caching policy we should use by default (unbounded memory versus weak-reference garbage collection if possible) is complex, but it probably should go more in #3.

As far as this particular issue goes (what function calls are considered equivalent), I definitely think that `this` and `new.target` should be distinguished in the cache.

**Comment by michaelficarra** [comment]
@js-choi Map keys don't use strict equality comparison, they use SameValueZero (and replace `-0` with `0`). So `NaN`s are equivalent, for example, unlike in strict equality comparison.

**Comment by js-choi** [comment]
@michaelficarra: Ah, yes, ugh, that is true…

**Comment by michaelficarra** [comment]
@js-choi If only Maps and Sets had used SameValue...

**Comment by js-choi** [comment]
@michaelficarra: Well, if we allow developers to supply custom user-supplied map-like caches, then the developer can supply a map-like object that treats `-0` specially:

```js
f.memo(new MapLikeCacheThatDistinguishesNegativeZero)
```

So the problem would be more about what *default* behavior we want. And, given that the Committee already decided that ordinary Maps should use SameValueZero, it likely will resolve on SameValueZero as the default memoization argument-matching behavior, too.

**Comment by zloirock** [comment]
> The tuple keys would have the structure `#[thisValue, newTargetValue, arg0, …]`

I'm not sure, what do you mean? Tuples can't store objects - only primitives, Records and other Tuples.

**Comment by NickGard** [comment]
Regardless of how the cache is built, the comparator function only really cares if the arguments for two different calls produce the same result. Or to get a little more in-depth, it cares if the _cache key_ produced by the arguments of two different calls are the same. It may be helpful to split up the idea of creating keys and comparing keys into different functions, say `hash` and `compare`. Then we could encode extra parameters outside of the argument list into the key like so:

```js
let extraParam = 2;

const memoFn = Function.memoize(fn, {
  hash: function (...args) {
    return [
      this,  // <- at call time, 'this' is captured as part of the cache key
      extraParam,  // <- at call time, extraParam is captured as part of the cache key
      ...args
    ];
  },
  compare: (keyA, keyB) => 
    keyA.length === keyB.length && 
    keyA.every((key, index) => key === keyB[index])
  )
});
```
